File: ios/DesMoinesInsider/Configuration/AppTypography.swift
import SwiftUI
import UIKit

// MARK: - App typography system
//
// Named text styles tuned for the app's hierarchy. Every style:
//
// - Scales with Dynamic Type, off the UIKit text style whose curve matches its
//   role, so "Larger Accessibility Sizes" actually enlarges the app
// - Applies a consistent weight + leading
// - Clamps at its own ceiling so a hero number can't push the layout apart
//
// The scaling is computed rather than declared. `Font.system(size:weight:)` is a
// FIXED point size - it ignores the user's text-size setting entirely - and for
// as long as this file used it, every one of these styles rendered at the same
// points at every Dynamic Type setting while the comment above claimed
// otherwise. `.dynamicTypeSize(...)` did not save it either: that clamps the
// environment for descendants, and a fixed font has nothing to clamp.
//
// Kept in parity with Android ui/theme/AppTypography.kt.

enum AppTextStyle {
    case displayLg       // Hero numbers, splash titles
    case displayMd       // Section hero titles
    case headline        // Primary page heading
    case title           // Card titles / dialog titles
    case subtitle        // Supporting title, secondary heading
    case body            // Primary body copy
    case bodyEmphasized  // Body with a heavier weight
    case bodySmall       // Detail / secondary body copy
    case caption         // Meta captions, timestamps
    case label           // All-caps labels / badge text
}

extension AppTextStyle {
    struct Spec {
        /// Point size at the default Dynamic Type setting (.large).
        let size: CGFloat
        let weight: Font.Weight
        /// The UIKit text style supplying the scaling curve. Headings grow more
        /// slowly than body copy, which is why this isn't one style for all ten.
        let metric: UIFont.TextStyle
        /// Ceiling past which this style stops growing.
        let maxTypeSize: DynamicTypeSize
        var tracking: CGFloat = 0
        var uppercased: Bool = false
    }

    var spec: Spec {
        switch self {
        case .displayLg:
            return Spec(size: 44, weight: .heavy, metric: .largeTitle,
                        maxTypeSize: .accessibility3, tracking: -0.5)
        case .displayMd:
            return Spec(size: 34, weight: .heavy, metric: .largeTitle,
                        maxTypeSize: .accessibility3, tracking: -0.3)
        case .headline:
            return Spec(size: 26, weight: .bold, metric: .title1,
                        maxTypeSize: .accessibility4)
        case .title:
            return Spec(size: 20, weight: .semibold, metric: .title3,
                        maxTypeSize: .accessibility5)
        case .subtitle:
            return Spec(size: 17, weight: .semibold, metric: .headline,
                        maxTypeSize: .accessibility5)
        case .body:
            return Spec(size: 16, weight: .regular, metric: .body,
                        maxTypeSize: .accessibility5)
        case .bodyEmphasized:
            return Spec(size: 16, weight: .semibold, metric: .body,
                        maxTypeSize: .accessibility5)
        case .bodySmall:
            return Spec(size: 14, weight: .regular, metric: .subheadline,
                        maxTypeSize: .accessibility5)
        case .caption:
            return Spec(size: 12, weight: .medium, metric: .caption1,
                        maxTypeSize: .accessibility5)
        case .label:
            return Spec(size: 11, weight: .bold, metric: .caption2,
                        maxTypeSize: .accessibility5, tracking: 0.8, uppercased: true)
        }
    }

    /// The point size this style renders at for `typeSize`, clamped at the
    /// style's own ceiling.
    func pointSize(for typeSize: DynamicTypeSize) -> CGFloat {
        let spec = spec
        let capped = min(typeSize, spec.maxTypeSize)
        return UIFontMetrics(forTextStyle: spec.metric).scaledValue(
            for: spec.size,
            compatibleWith: UITraitCollection(preferredContentSizeCategory: capped.contentSizeCategory)
        )
    }
}

extension DynamicTypeSize {
    /// UIKit equivalent, needed because `UIFontMetrics` works in content size
    /// categories and SwiftUI hands us a `DynamicTypeSize`.
    var contentSizeCategory: UIContentSizeCategory {
        switch self {
        case .xSmall: return .extraSmall
        case .small: return .small
        case .medium: return .medium
        case .large: return .large
        case .xLarge: return .extraLarge
        case .xxLarge: return .extraExtraLarge
        case .xxxLarge: return .extraExtraExtraLarge
        case .accessibility1: return .accessibilityMedium
        case .accessibility2: return .accessibilityLarge
        case .accessibility3: return .accessibilityExtraLarge
        case .accessibility4: return .accessibilityExtraExtraLarge
        case .accessibility5: return .accessibilityExtraExtraExtraLarge
        @unknown default: return .large
        }
    }
}

extension View {
    func appText(_ style: AppTextStyle) -> some View {
        modifier(AppTextStyleModifier(style: style))
    }
}

private struct AppTextStyleModifier: ViewModifier {
    let style: AppTextStyle

    @Environment(\.dynamicTypeSize) private var typeSize

    func body(content: Content) -> some View {
        let spec = style.spec
        return content
            .font(.system(size: style.pointSize(for: typeSize), weight: spec.weight))
            .tracking(spec.tracking)
            .textCase(spec.uppercased ? Text.Case.uppercase : nil)
            // Descendants that set their own font stay inside this style's
            // ceiling too; the font above is already clamped by pointSize().
            .dynamicTypeSize(...spec.maxTypeSize)
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/ui/theme/AppTypography.kt
package com.desmoines.aipulse.ui.theme

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.sp

/**
 * App text style tokens kept in parity with iOS AppTypography.swift.
 *
 * Named styles for the app's hierarchy — every style has a consistent weight,
 * letter-spacing, and leading. Backed by Material 3's Typography system so
 * Compose previews and theming integrations still "just work".
 */
enum class AppTextStyle {
    DisplayLg,       // Hero numbers, splash titles
    DisplayMd,       // Section hero titles
    Headline,        // Primary page heading
    Title,           // Card titles, dialog titles
    Subtitle,        // Supporting title, secondary heading
    Body,            // Primary body copy
    BodyEmphasized,  // Body with a heavier weight
    BodySmall,       // Detail / secondary body copy
    Caption,         // Meta captions, timestamps
    Label,           // All-caps labels / badge text
}

private fun AppTextStyle.toTextStyle(base: TextStyle): TextStyle = when (this) {
    AppTextStyle.DisplayLg -> base.copy(fontSize = 44.sp, fontWeight = FontWeight.Black, letterSpacing = (-0.5).sp, lineHeight = 48.sp)
    AppTextStyle.DisplayMd -> base.copy(fontSize = 34.sp, fontWeight = FontWeight.Black, letterSpacing = (-0.3).sp, lineHeight = 40.sp)
    AppTextStyle.Headline -> base.copy(fontSize = 26.sp, fontWeight = FontWeight.Bold, lineHeight = 32.sp)
    AppTextStyle.Title -> base.copy(fontSize = 20.sp, fontWeight = FontWeight.SemiBold, lineHeight = 26.sp)
    AppTextStyle.Subtitle -> base.copy(fontSize = 17.sp, fontWeight = FontWeight.SemiBold, lineHeight = 22.sp)
    AppTextStyle.Body -> base.copy(fontSize = 16.sp, fontWeight = FontWeight.Normal, lineHeight = 22.sp)
    AppTextStyle.BodyEmphasized -> base.copy(fontSize = 16.sp, fontWeight = FontWeight.SemiBold, lineHeight = 22.sp)
    AppTextStyle.BodySmall -> base.copy(fontSize = 14.sp, fontWeight = FontWeight.Normal, lineHeight = 18.sp)
    AppTextStyle.Caption -> base.copy(fontSize = 12.sp, fontWeight = FontWeight.Medium, lineHeight = 15.sp)
    AppTextStyle.Label -> base.copy(fontSize = 11.sp, fontWeight = FontWeight.Bold, letterSpacing = 0.8.sp, lineHeight = 14.sp)
}

/**
 * Composable helper that renders a [Text] using the app's typography token.
 * Use when you want to avoid manually copying font sizes / weights into
 * feature code.
 */
@Composable
fun AppText(
    text: String,
    style: AppTextStyle,
    modifier: Modifier = Modifier,
    color: Color = Color.Unspecified,
    maxLines: Int = Int.MAX_VALUE,
    overflow: TextOverflow = TextOverflow.Clip,
) {
    val base = MaterialTheme.typography.bodyLarge
    Text(
        text = if (style == AppTextStyle.Label) text.uppercase() else text,
        modifier = modifier,
        style = style.toTextStyle(base),
        color = color,
        maxLines = maxLines,
        overflow = overflow,
    )
}

/** Resolve an [AppTextStyle] to a raw [TextStyle] for use with e.g. custom components. */
@Composable
fun rememberAppTextStyle(style: AppTextStyle): TextStyle {
    val base = MaterialTheme.typography.bodyLarge
    return style.toTextStyle(base)
}
